url7f15a160d2_timofei7_old_ai_assignments_tgz-pJ8-org_json_TestJ8/src/hw4/CSP.java

```diff
@@ -0,0 +1,496 @@
+				{
+					size = domainlist.getValues(v).size();
+					variable = v;
+				}
+			}
+			PriorityQueue<priorInt> order = new PriorityQueue<priorInt>();
+			
+			int o = 0;
+			for(Integer value : domainlist.getValues(variable))
+			{
+				PartialAssignment ta = (PartialAssignment) assign.clone();
+				ta.set(variable, value);
+				order.add(new priorInt(value,ta.getUnassignedVariables().size()));
+				o++;
+			}
+			//ordered by least constraining value
+			for(int i=1; i < domainlist.getValues(variable).size(); i++)
+			{
+				if (!solutionFound.value)
+				{
+					Integer value = order.poll().value;
+					pa.set(variable, value);
+					//DomainList newdomainlist = forwardCheck(variable, value, pa, domainlist);
+					solutionFound = backtrackingSearchMRVLCV(pa, domainlist);
+				}
+			}
+			return solutionFound;
+		}
+	}
+		
+	/**
+	 * whenever a variable x is assigned, establish arc-consistency:
+	 * for each unassigned variable y that is connected to x by a constraint
+	 * delete from y's domain any value that is inconsistent with the value chosen for X
+	 * @param curr_variable
+	 * @param pa
+	 * @param dl
+	 * @return
+	 */
+	public DomainList forwardCheck(Integer x, Integer chosenValue, final PartialAssignment pa, final DomainList dl)
+	{
+		DomainList ndl = (DomainList) dl.clone();
+		for (Integer y : pa.getUnassignedVariables()) //for each unassigned variable y
+		{
+			IntegerPair xy = new IntegerPair(x,y);
+			if (y != x && constrainedVariablesList.contains(xy)) //that is connected to x by a constraint and isn't x itself
+			{
+				for (Integer value : dl.getValues(y)) //go over all values in y's domain
+				{
+					if (!constraintTable.get(xy).satisfied(chosenValue, value)) //if value is inconsistent with the one chosen for x
+					{
+						//System.out.println("chosenValue: " +chosenValue + " value:" + value);
+						//System.out.println(constraintTable.get(xy).toString());
+						//System.out.println("deleting: " +value +" from: " + variableNames.get(y));
+						ndl.deleteValue(y, value); //delete value from y's domain
+						//System.out.println(ndl.getValues(y).toString());
+					}
+				}
+			}
+		}
+		
+		return ndl;
+	}
+	
+	
+	/**
+	 * converts to conjunctive normal form
+	 * in this case i've generalized it so they appear as their internal id number rather than the string values
+	 * TODO: this doesn't really work, prints DNF for now as it is waaay shorter
+	 */
+	public void outputCNF()
+	{
+		System.out.println("CNF: (well I lied...right now actually in DNF...)");
+		for (IntegerPair ip: constrainedVariablesList)
+		{
+			Constraint c = constraintTable.get(ip);
+			String or = "";
+			//LinkedList<String[]> temp = new LinkedList<String[]>();
+			for (IntegerPair p : c.allowedPairs)
+			{   
+				//temp.add(new String[] {variableNames.get(ip.first)+p.first, variableNames.get(ip.second)+p.second});
+				if (or == "") or = "("+variableNames.get(ip.first)+p.first+" ^ "+variableNames.get(ip.second)+p.second+")";
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.